### ISSUE TITLE
Forced EOS token in vllm generation?

### ISSUE THREAD

**Issue by mgerstgrasser** [OpenRLHF/OpenRLHF#238]
I see in `RemoteExperienceMaker._generate_vllm()`, [line 375](https://github.com/OpenLLMAI/OpenRLHF/blob/4e15591a5abd19a14e4a72415603bce76c3e1567/openrlhf/trainer/ppo_utils/experience_maker.py#L375) that for generations that don't finish, i.e. don't output the EOS tokens within the max token limit, we manually set the last token to be the EOS token, even though that was not what the model generated.

Isn't this the wrong thing to do? E.g. if the model generated an unfinished sentence like "This is an unfinished" when it ran into the token limit, shouldn't we train on that, rather than "This is an </s>"? My understanding of the PPO algorithm is also that it doesn't do well with off-policy experiences, which we technically have if we manually change to the EOS token. So I just wanted to check if there's a specific reason to do this?

It also looks to me that the huggingface `model.generate()` method, and by extension `RemoteExperienceMaker._generate_local()` and `NaiveExperienceMaker` do not do this.

**Comment by hijkzzz** [comment]
Because the reward model uses EOS token to predict the reward value

**Comment by mgerstgrasser** [comment]
Ahhhh, got it, that makes sense. I think that's probably broken with local generation then! I just verified that that doesn't have EOS if max_tokens is reached.

Also, would taking the RM output on the last non-masked token instead of EOS be a better way around this? Or alternatively, forcing the EOS token only when feeding the experience to the RM?

**Comment by hijkzzz** [comment]
I am not sure which approach to take at the moment, but our current implementation is heavily dependent on EOS tokens.

**Comment by mgerstgrasser** [comment]
> I am not sure which approach to take at the moment, but our current implementation is heavily dependent on EOS tokens.

You mean specifically for the RM? Or more broadly than that?

**Comment by mgerstgrasser** [comment]
Oh, for local generation, does `actor.process_sequences()` do the same thing? https://github.com/OpenLLMAI/OpenRLHF/blob/bed10e115523a9eca419cb058ede8e531d23c182/openrlhf/models/actor.py#L159

If so, then doing this in `RemoteExperienceMaker` seems unnecessary, since that also calls `actor.process_sequences()` later anyway, i.e. this is being done twice, I think?

**Comment by mgerstgrasser** [comment]
@hijkzzz Could I ask a quick related question: In `actor.process_sequences()` I also see that `attention_mask` is set to False on all EOS tokens, except the final EOS token in each sequence. In the datasets used in the examples, it seems that there never is an EOS token in the input, but if there was (e.g. in a previous turn in a conversation), shouldn't the attention mask be True there?

**Comment by pagepal666** [comment]
> @hijkzzz Could I ask a quick related question: In `actor.process_sequences()` I also see that `attention_mask` is set to False on all EOS tokens, except the final EOS token in each sequence. In the datasets used in the examples, it seems that there never is an EOS token in the input, but if there was (e.g. in a previous turn in a conversation), shouldn't the attention mask be True there?

You need to change to a new special token

**Comment by Silent-Zebra** [comment]
I agree with @mgerstgrasser that the forced EOS token that replaces the last token generation is wrong. I think it causes bugs with calculating KL divergence (because the KL divergence assumes you have samples which come from your current model, which is not the case when the actual model generation is overwritten in the last token spot; thus your estimated KL can even be very negative), which can then cause the reward to be wrong, and therefore also the value and policy to be wrong. I believe this issue is reproducible by setting --generate_max_len to a small number such as 2. 

I guess the reason most people usually wouldn't see much of an effect from this issue is that most people probably use long sequence lengths, and as a result most sequences probably generate an EOS token before the end anyway, so these sequences are unaffected. But I believe the issue is more problematic the smaller the number of tokens you generate.

Solutions: As pointed out by @mgerstgrasser, we could force the EOS token into only the reward model (and perhaps concatenate instead of overwriting the last token), if this is the only place where the final EOS is needed. If the final EOS is needed elsewhere, we could concatenate an additional final token to all generations, that is either EOS or padding for sequences that have not yet generated EOS. Perhaps an easy way of doing this is just concatenating a final padding token to all tokens before running the process_sequences() function? (Not sure if having an extra token messes up other things in the pipeline)

Currently I am using custom models, so I fixed this issue by simply just cutting out the lines in process_sequences() that use the EOS token (without doing this, I was getting KL values like -40 or so, and had very unstable training and nonsensical policies). If I get around to testing with the existing supported models, and come up with a fix that works for those (e.g. along the lines of the suggested strategies above), I will follow up. Otherwise, if any of you fix this issue, I would greatly appreciate it. I value toy settings a lot for the purposes of understanding, debugging, and fast iteration, and think it would be valuable to the broader community to address this issue.
